File: swift-client/Sources/ParRtDbClient/QueryCombinations.swift
// GENERATED transcription of wire-corpus/query-combinations.json — kept byte-parity
// with the JSON source via QueryCombinationsParityTests. Do not hand-edit without
// re-running the transcription; a mismatch fails that test loudly.

/// One rule from `wire-corpus/query-combinations.json`: either `forbid` (all listed
/// clauses present is an error) or `atMostOne` (more than one listed clause present is
/// an error) — never both.
struct QueryCombinationRule: Equatable, Codable, Sendable {
    let id: String
    let forbid: [String]?
    let atMostOne: [String]?
    let code: String
    let message: String
}

/// The table-driven replacement for the hand-written combination ladder (ENH-028
/// phase 2) — the swift mirror of `wire-corpus/query-combinations.json`, which every
/// runner (server + all four client in-memory engines) now shares as the single source
/// of truth for which read-`Query` clauses may not be set together.
enum QueryCombinationRules {
    /// Every clause name a rule can reference (wire camelCase).
    static let clauses: [String] = [
        "get", "index", "eq", "gt", "gte", "lt", "lte", "order", "take", "unique", "first",
        "count", "distinct", "aggregate", "paginate", "filter", "search", "vectorSearch",
        "hybridSearch"
    ]

    /// Rules in table order — order has no effect on which `code` a query fails with
    /// (every rule here is `BAD_REQUEST`), but is preserved for parity with the JSON.
    static let rules: [QueryCombinationRule] = [
        QueryCombinationRule(
            id: "terminal-exclusive",
            forbid: nil,
            atMostOne: ["aggregate", "count", "distinct", "first", "get", "paginate", "take", "unique"],
            code: "BAD_REQUEST",
            message: "only one terminal may be set"
        ),
        QueryCombinationRule(
            id: "search-mode-exclusive",
            forbid: nil,
            atMostOne: ["hybridSearch", "search", "vectorSearch"],
            code: "BAD_REQUEST",
            message: "only one search mode terminal may be set"
        ),
        QueryCombinationRule(
            id: "get-excludes-index",
            forbid: ["get", "index"],
            atMostOne: nil,
            code: "BAD_REQUEST",
            message: "get cannot be combined with index, eq, range bounds, order, take, unique, first, count, "
                + "distinct, aggregate, paginate, filter, search, or vector search"
        ),
        QueryCombinationRule(
            id: "eq-excludes-get",
            forbid: ["eq", "get"],
            atMostOne: nil,
            code: "BAD_REQUEST",
            message: "get cannot be combined with index, eq, range bounds, order, take, unique, first, count, "
                + "distinct, aggregate, paginate, filter, search, or vector search"
        ),
        QueryCombinationRule(
            id: "get-excludes-gt",
            forbid: ["get", "gt"],
            atMostOne: nil,
            code: "BAD_REQUEST",
            message: "get cannot be combined with index, eq, range bounds, order, take, unique, first, count, "
                + "distinct, aggregate, paginate, filter, search, or vector search"
        ),
        QueryCombinationRule(
            id: "get-excludes-gte",
            forbid: ["get", "gte"],
            atMostOne: nil,
            code: "BAD_REQUEST",
            message: "get cannot be combined with index, eq, range bounds, order, take, unique, first, count, "
                + "distinct, aggregate, paginate, filter, search, or vector search"
        ),
        QueryCombinationRule(
            id: "get-excludes-lt",
            forbid: ["get", "lt"],
            atMostOne: nil,
            code: "BAD_REQUEST",
            message: "get cannot be combined with index, eq, range bounds, order, take, unique, first, count, "
                + "distinct, aggregate, paginate, filter, search, or vector search"
        ),
        QueryCombinationRule(
            id: "get-excludes-lte",
            forbid: ["get", "lte"],
            atMostOne: nil,
            code: "BAD_REQUEST",
            message: "get cannot be combined with index, eq, range bounds, order, take, unique, first, count, "
                + "distinct, aggregate, paginate, filter, search, or vector search"
        ),
        QueryCombinationRule(
            id: "get-excludes-order",
            forbid: ["get", "order"],
            atMostOne: nil,
            code: "BAD_REQUEST",
            message: "get cannot be combined with index, eq, range bounds, order, take, unique, first, count, "
                + "distinct, aggregate, paginate, filter, search, or vector search"
        ),
        QueryCombinationRule(
            id: "filter-excludes-get",
            forbid: ["filter", "get"],
            atMostOne: nil,
            code: "BAD_REQUEST",
            message: "get cannot be combined with index, eq, range bounds, order, take, unique, first, count, "
                + "distinct, aggregate, paginate, filter, search, or vector search"
        ),
        QueryCombinationRule(
            id: "get-excludes-search",
            forbid: ["get", "search"],
            atMostOne: nil,
            code: "BAD_REQUEST",
            message: "get cannot be combined with index, eq, range bounds, order, take, unique, first, count, "
                + "distinct, aggregate, paginate, filter, search, or vector search"
        ),
        QueryCombinationRule(
            id: "get-excludes-vectorSearch",
            forbid: ["get", "vectorSearch"],
            atMostOne: nil,
            code: "BAD_REQUEST",
            message: "get cannot be combined with index, eq, range bounds, order, take, unique, first, count, "
                + "distinct, aggregate, paginate, filter, search, or vector search"
        ),
        QueryCombinationRule(
            id: "get-excludes-hybridSearch",
            forbid: ["get", "hybridSearch"],
            atMostOne: nil,
            code: "BAD_REQUEST",
            message: "get cannot be combined with index, eq, range bounds, order, take, unique, first, count, "
                + "distinct, aggregate, paginate, filter, search, or vector search"
        ),
        QueryCombinationRule(
            id: "order-excludes-unique",
            forbid: ["order", "unique"],
            atMostOne: nil,
            code: "BAD_REQUEST",
            message: "unique cannot be combined with take, order, distinct, or aggregate"
        ),
        QueryCombinationRule(
            id: "count-excludes-order",
            forbid: ["count", "order"],
            atMostOne: nil,
            code: "BAD_REQUEST",
            message: "count cannot be combined with order"
        ),
        QueryCombinationRule(
            id: "distinct-excludes-order",
            forbid: ["distinct", "order"],
            atMostOne: nil,
            code: "BAD_REQUEST",
            message: "distinct cannot be combined with order"
        ),
        QueryCombinationRule(
            id: "distinct-excludes-search",
            forbid: ["distinct", "search"],
            atMostOne: nil,
            code: "BAD_REQUEST",
            message: "distinct cannot be combined with search"
        ),
        QueryCombinationRule(
            id: "distinct-excludes-vectorSearch",
            forbid: ["distinct", "vectorSearch"],
            atMostOne: nil,
            code: "BAD_REQUEST",
            message: "distinct cannot be combined with vector search"
        ),
        QueryCombinationRule(
            id: "distinct-excludes-hybridSearch",
            forbid: ["distinct", "hybridSearch"],
            atMostOne: nil,
            code: "BAD_REQUEST",
            message: "distinct cannot be combined with hybrid search"
        ),
        QueryCombinationRule(
            id: "aggregate-excludes-order",
            forbid: ["aggregate", "order"],
            atMostOne: nil,
            code: "BAD_REQUEST",
            message: "aggregate cannot be combined with order"
        ),
        QueryCombinationRule(
            id: "aggregate-excludes-search",
            forbid: ["aggregate", "search"],
            atMostOne: nil,
            code: "BAD_REQUEST",
            message: "aggregate cannot be combined with search"
        ),
        QueryCombinationRule(
            id: "aggregate-excludes-vectorSearch",
            forbid: ["aggregate", "vectorSearch"],
            atMostOne: nil,
            code: "BAD_REQUEST",
            message: "aggregate cannot be combined with vector search"
        ),
        QueryCombinationRule(
            id: "aggregate-excludes-hybridSearch",
            forbid: ["aggregate", "hybridSearch"],
            atMostOne: nil,
            code: "BAD_REQUEST",
            message: "aggregate cannot be combined with hybrid search"
        ),
        QueryCombinationRule(
            id: "gt-excludes-gte",
            forbid: ["gt", "gte"],
            atMostOne: nil,
            code: "BAD_REQUEST",
            message: "gt and gte cannot both be set"
        ),
        QueryCombinationRule(
            id: "lt-excludes-lte",
            forbid: ["lt", "lte"],
            atMostOne: nil,
            code: "BAD_REQUEST",
            message: "lt and lte cannot both be set"
        ),
        QueryCombinationRule(
            id: "index-excludes-vectorSearch",
            forbid: ["index", "vectorSearch"],
            atMostOne: nil,
            code: "BAD_REQUEST",
            message: "vectorSearch cannot be combined with any other terminal"
        ),
        QueryCombinationRule(
            id: "eq-excludes-vectorSearch",
            forbid: ["eq", "vectorSearch"],
            atMostOne: nil,
            code: "BAD_REQUEST",
            message: "vectorSearch cannot be combined with any other terminal"
        ),
        QueryCombinationRule(
            id: "gt-excludes-vectorSearch",
            forbid: ["gt", "vectorSearch"],
            atMostOne: nil,
            code: "BAD_REQUEST",
            message: "vectorSearch cannot be combined with any other terminal"
        ),
        QueryCombinationRule(
            id: "gte-excludes-vectorSearch",
            forbid: ["gte", "vectorSearch"],
            atMostOne: nil,
            code: "BAD_REQUEST",
            message: "vectorSearch cannot be combined with any other terminal"
        ),
        QueryCombinationRule(
            id: "lt-excludes-vectorSearch",
            forbid: ["lt", "vectorSearch"],
            atMostOne: nil,
            code: "BAD_REQUEST",
            message: "vectorSearch cannot be combined with any other terminal"
        ),
        QueryCombinationRule(
            id: "lte-excludes-vectorSearch",
            forbid: ["lte", "vectorSearch"],
            atMostOne: nil,
            code: "BAD_REQUEST",
            message: "vectorSearch cannot be combined with any other terminal"
        ),
        QueryCombinationRule(
            id: "order-excludes-vectorSearch",
            forbid: ["order", "vectorSearch"],
            atMostOne: nil,
            code: "BAD_REQUEST",
            message: "vectorSearch cannot be combined with any other terminal"
        ),
        QueryCombinationRule(
            id: "unique-excludes-vectorSearch",
            forbid: ["unique", "vectorSearch"],
            atMostOne: nil,
            code: "BAD_REQUEST",
            message: "vectorSearch cannot be combined with any other terminal"
        ),
        QueryCombinationRule(
            id: "first-excludes-vectorSearch",
            forbid: ["first", "vectorSearch"],
            atMostOne: nil,
            code: "BAD_REQUEST",
            message: "vectorSearch cannot be combined with any other terminal"
        ),
        QueryCombinationRule(
            id: "count-excludes-vectorSearch",
            forbid: ["count", "vectorSearch"],
            atMostOne: nil,
            code: "BAD_REQUEST",
            message: "vectorSearch cannot be combined with any other terminal"
        ),
        QueryCombinationRule(
            id: "paginate-excludes-vectorSearch",
            forbid: ["paginate", "vectorSearch"],
            atMostOne: nil,
            code: "BAD_REQUEST",
            message: "vectorSearch cannot be combined with any other terminal"
        ),
        QueryCombinationRule(
            id: "filter-excludes-vectorSearch",
            forbid: ["filter", "vectorSearch"],
            atMostOne: nil,
            code: "BAD_REQUEST",
            message: "vectorSearch cannot be combined with any other terminal"
        ),
        QueryCombinationRule(
            id: "take-excludes-vectorSearch",
            forbid: ["take", "vectorSearch"],
            atMostOne: nil,
            code: "BAD_REQUEST",
            message: "vectorSearch cannot be combined with any other terminal"
        ),
        QueryCombinationRule(
            id: "hybridSearch-excludes-index",
            forbid: ["hybridSearch", "index"],
            atMostOne: nil,
            code: "BAD_REQUEST",
            message: "hybridSearch cannot be combined with any other terminal"
        ),
        QueryCombinationRule(
            id: "eq-excludes-hybridSearch",
            forbid: ["eq", "hybridSearch"],
            atMostOne: nil,
            code: "BAD_REQUEST",
            message: "hybridSearch cannot be combined with any other terminal"
        ),
        QueryCombinationRule(
            id: "gt-excludes-hybridSearch",
            forbid: ["gt", "hybridSearch"],
            atMostOne: nil,
            code: "BAD_REQUEST",
            message: "hybridSearch cannot be combined with any other terminal"
        ),
        QueryCombinationRule(
            id: "gte-excludes-hybridSearch",
            forbid: ["gte", "hybridSearch"],
            atMostOne: nil,
            code: "BAD_REQUEST",
            message: "hybridSearch cannot be combined with any other terminal"
        ),
        QueryCombinationRule(
            id: "hybridSearch-excludes-lt",
            forbid: ["hybridSearch", "lt"],
            atMostOne: nil,
            code: "BAD_REQUEST",
            message: "hybridSearch cannot be combined with any other terminal"
        ),
        QueryCombinationRule(
            id: "hybridSearch-excludes-lte",
            forbid: ["hybridSearch", "lte"],
            atMostOne: nil,
            code: "BAD_REQUEST",
            message: "hybridSearch cannot be combined with any other terminal"
        ),
        QueryCombinationRule(
            id: "hybridSearch-excludes-order",
            forbid: ["hybridSearch", "order"],
            atMostOne: nil,
            code: "BAD_REQUEST",
            message: "hybridSearch cannot be combined with any other terminal"
        ),
        QueryCombinationRule(
            id: "hybridSearch-excludes-take",
            forbid: ["hybridSearch", "take"],
            atMostOne: nil,
            code: "BAD_REQUEST",
            message: "hybridSearch cannot be combined with any other terminal"
        ),
        QueryCombinationRule(
            id: "hybridSearch-excludes-unique",
            forbid: ["hybridSearch", "unique"],
            atMostOne: nil,
            code: "BAD_REQUEST",
            message: "hybridSearch cannot be combined with any other terminal"
        ),
        QueryCombinationRule(
            id: "first-excludes-hybridSearch",
            forbid: ["first", "hybridSearch"],
            atMostOne: nil,
            code: "BAD_REQUEST",
            message: "hybridSearch cannot be combined with any other terminal"
        ),
        QueryCombinationRule(
            id: "count-excludes-hybridSearch",
            forbid: ["count", "hybridSearch"],
            atMostOne: nil,
            code: "BAD_REQUEST",
            message: "hybridSearch cannot be combined with any other terminal"
        ),
        QueryCombinationRule(
            id: "hybridSearch-excludes-paginate",
            forbid: ["hybridSearch", "paginate"],
            atMostOne: nil,
            code: "BAD_REQUEST",
            message: "hybridSearch cannot be combined with any other terminal"
        ),
        QueryCombinationRule(
            id: "filter-excludes-hybridSearch",
            forbid: ["filter", "hybridSearch"],
            atMostOne: nil,
            code: "BAD_REQUEST",
            message: "hybridSearch cannot be combined with any other terminal"
        ),
        QueryCombinationRule(
            id: "index-excludes-search",
            forbid: ["index", "search"],
            atMostOne: nil,
            code: "BAD_REQUEST",
            message: "search cannot be combined with index, eq, range bounds, order, unique, first, count, distinct, "
                + "aggregate, paginate, filter, or vector search"
        ),
        QueryCombinationRule(
            id: "eq-excludes-search",
            forbid: ["eq", "search"],
            atMostOne: nil,
            code: "BAD_REQUEST",
            message: "search cannot be combined with index, eq, range bounds, order, unique, first, count, distinct, "
                + "aggregate, paginate, filter, or vector search"
        ),
        QueryCombinationRule(
            id: "gt-excludes-search",
            forbid: ["gt", "search"],
            atMostOne: nil,
            code: "BAD_REQUEST",
            message: "search cannot be combined with index, eq, range bounds, order, unique, first, count, distinct, "
                + "aggregate, paginate, filter, or vector search"
        ),
        QueryCombinationRule(
            id: "gte-excludes-search",
            forbid: ["gte", "search"],
            atMostOne: nil,
            code: "BAD_REQUEST",
            message: "search cannot be combined with index, eq, range bounds, order, unique, first, count, distinct, "
                + "aggregate, paginate, filter, or vector search"
        ),
        QueryCombinationRule(
            id: "lt-excludes-search",
            forbid: ["lt", "search"],
            atMostOne: nil,
            code: "BAD_REQUEST",
            message: "search cannot be combined with index, eq, range bounds, order, unique, first, count, distinct, "
                + "aggregate, paginate, filter, or vector search"
        ),
        QueryCombinationRule(
            id: "lte-excludes-search",
            forbid: ["lte", "search"],
            atMostOne: nil,
            code: "BAD_REQUEST",
            message: "search cannot be combined with index, eq, range bounds, order, unique, first, count, distinct, "
                + "aggregate, paginate, filter, or vector search"
        ),
        QueryCombinationRule(
            id: "order-excludes-search",
            forbid: ["order", "search"],
            atMostOne: nil,
            code: "BAD_REQUEST",
            message: "search cannot be combined with index, eq, range bounds, order, unique, first, count, distinct, "
                + "aggregate, paginate, filter, or vector search"
        ),
        QueryCombinationRule(
            id: "search-excludes-unique",
            forbid: ["search", "unique"],
            atMostOne: nil,
            code: "BAD_REQUEST",
            message: "search cannot be combined with index, eq, range bounds, order, unique, first, count, distinct, "
                + "aggregate, paginate, filter, or vector search"
        ),
        QueryCombinationRule(
            id: "first-excludes-search",
            forbid: ["first", "search"],
            atMostOne: nil,
            code: "BAD_REQUEST",
            message: "search cannot be combined with index, eq, range bounds, order, unique, first, count, distinct, "
                + "aggregate, paginate, filter, or vector search"
        ),
        QueryCombinationRule(
            id: "count-excludes-search",
            forbid: ["count", "search"],
            atMostOne: nil,
            code: "BAD_REQUEST",
            message: "search cannot be combined with index, eq, range bounds, order, unique, first, count, distinct, "
                + "aggregate, paginate, filter, or vector search"
        ),
        QueryCombinationRule(
            id: "paginate-excludes-search",
            forbid: ["paginate", "search"],
            atMostOne: nil,
            code: "BAD_REQUEST",
            message: "search cannot be combined with index, eq, range bounds, order, unique, first, count, distinct, "
                + "aggregate, paginate, filter, or vector search"
        ),
        QueryCombinationRule(
            id: "filter-excludes-search",
            forbid: ["filter", "search"],
            atMostOne: nil,
            code: "BAD_REQUEST",
            message: "search cannot be combined with index, eq, range bounds, order, unique, first, count, distinct, "
                + "aggregate, paginate, filter, or vector search"
        )
    ]
}

/// Returns the first rule (in table order) violated by the given set of present clause
/// names, or `nil` if the combination is legal.
func firstViolatedQueryCombinationRule(_ present: Set<String>) -> QueryCombinationRule? {
    for rule in QueryCombinationRules.rules {
        if let forbid = rule.forbid {
            if forbid.allSatisfy({ present.contains($0) }) {
                return rule
            }
        } else if let atMostOne = rule.atMostOne {
            let presentCount = atMostOne.count { present.contains($0) }
            if presentCount > 1 {
                return rule
            }
        }
    }
    return nil
}

File: core/src/query_combinations.rs
//! Shared read-`Query` combination-rule evaluator (ENH-028 phase 2). Phase 1
//! (`wire-corpus/query-combinations.json` + one `wire-corpus/semantics/
//! query-combo-<id>.json` corpus case per rule) found zero behavioral drift
//! between the server's hand-written cascade (`server/src/query/mod.rs`'s
//! `GET_PEERS`/`UNIQUE_PEERS`/`*_INCOMPATIBLES`/`*_PEERS` tables, consumed by
//! `terminals.rs::compile_query`) and the Rust client's hand-written
//! `check_query_combinations` — this module replaces both with one
//! table-driven evaluator neither crate has to hand-maintain.
//!
//! The table has two rule shapes: `forbid` (every listed clause present at
//! once is rejected) and `atMostOne` (more than one listed clause present at
//! once is rejected — used only where the listed set is a full pairwise
//! clique, so it never wrongly forbids two clauses that ARE legal together).
//! `atMostOne` cliques (`terminal-exclusive`, `search-mode-exclusive`) are
//! phase 1 CONSOLIDATIONS of what used to be many individually-worded
//! pairwise checks (e.g. the pre-refactor `AGGREGATE_INCOMPATIBLES` table's
//! per-peer `"aggregate cannot be combined with X"` messages), so evaluation
//! checks every `forbid` rule (in declared order) before any `atMostOne`
//! rule (in declared order): a `forbid` rule's message is always at least as
//! specific as a clique's generic one, so it wins when both fire on the same
//! query. This tie-break only decides which `message` a multi-violation query
//! gets — accept/reject is a boolean OR over every rule regardless of
//! evaluation order — and the corpus asserts `code` only, per
//! `wire-corpus/README.md`'s determinism ruling 4, so it is not part of the
//! enforced contract. Every declared rule shares `code: "BAD_REQUEST"` today.
//!
//! Known gap (ENH-028 phase 2, filed against phase 1's table, NOT fixed
//! here): most pairs inside the `terminal-exclusive` clique
//! (`aggregate`/`count`/`distinct`/`first`/`get`/`paginate`/`take`/`unique`)
//! have no matching `forbid` rule in `wire-corpus/query-combinations.json`
//! and so fall through to the clique's generic `"only one terminal may be
//! set"` message instead of the pre-refactor per-peer wording — a
//! `code`-level no-op (still `BAD_REQUEST`) but a message regression.
//! Confirmed (2026-08-25, via a scratch probe against this evaluator, not
//! committed) missing pairs: `get`+`unique`/`distinct`/`aggregate`,
//! `unique`+`take`, `first`+`unique`/`take`, `count`+`unique`/`take`/`first`,
//! `aggregate`+`take`, `distinct`+`take`, `paginate`+`take`/`count`/`unique`/
//! `first` — only the `*-excludes-order` pairs (`order-excludes-unique`,
//! `count-excludes-order`, `distinct-excludes-order`,
//! `aggregate-excludes-order`) got an individual rule; every other
//! terminal-vs-terminal pair relies solely on the clique. This breaks four
//! pre-existing `rust-client` unit tests that assert the old per-peer
//! wording (`in_memory::tests::{aggregate::{aggregate,distinct}_rejects_conflicting_terminals,
//! paginate::paginate_rejects_combination_with_take_count_unique_or_first,
//! query::query_rejects_conflicting_terminals}`). Closing it means adding a
//! `forbid` rule per missing pair (or restoring per-peer messages some other
//! way) plus a `wire-corpus/semantics/query-combo-<id>.json` case per new
//! rule id (see `server/src/query/combinations_coverage.rs`'s coverage
//! guard) — out of this phase's touch scope:
//! `wire-corpus/query-combinations.json` is a cross-runner file four other
//! language agents are consuming concurrently.

use std::collections::HashSet;
use std::sync::LazyLock;

use serde::Deserialize;

/// `wire-corpus/query-combinations.json`, embedded at compile time so neither
/// crate needs a runtime file read (and so a missing/relocated file is a
/// build error, not a startup surprise).
static RULES_JSON: &str = include_str!("../../wire-corpus/query-combinations.json");

#[derive(Debug, Deserialize)]
struct RuleTable {
    rules: Vec<Rule>,
}

#[derive(Debug, Deserialize)]
struct Rule {
    #[serde(default)]
    forbid: Option<Vec<String>>,
    #[serde(default, rename = "atMostOne")]
    at_most_one: Option<Vec<String>>,
    code: String,
    message: String,
}

static RULES: LazyLock<RuleTable> = LazyLock::new(|| {
    let table: RuleTable = serde_json::from_str(RULES_JSON)
        .expect("wire-corpus/query-combinations.json must parse as a RuleTable");
    for rule in &table.rules {
        assert!(
            rule.forbid.is_some() != rule.at_most_one.is_some(),
            "wire-corpus/query-combinations.json: a rule must declare exactly one of \
             forbid/atMostOne"
        );
    }
    table
});

/// One violated query-combination rule. `code` mirrors the JSON table's
/// `code` (currently always `"BAD_REQUEST"`); `message` is the table's
/// canonical wording. This is a plain data seam, not either crate's error
/// type — `server`'s `RtDbError` and `rust-client`'s `RtDbError` are separate,
/// non-`core` types, so each caller translates a `RuleViolation` into its own
/// error at the call site, the same seam [`crate::engine::detect_destructive_changes`]
/// already uses for its `Result<(), String>`.
#[derive(Debug, Clone)]
pub struct RuleViolation {
    /// The rule's error code (currently always `"BAD_REQUEST"`).
    pub code: String,
    /// The rule's canonical human-readable message.
    pub message: String,
}

/// Evaluate every declared query-combination rule against `present` — the
/// wire-corpus canonical clause names (see the JSON table's `clauses` array,
/// e.g. `"get"`, `"index"`, `"vectorSearch"`) for every clause the query
/// actually sets — and return the first violated rule, or `Ok(())` if none
/// fire. Callers build `present` once per query (map each optional/boolean
/// `Query` field to its clause name) and call this once, replacing the
/// server's per-terminal peer-rejection cascade and the Rust client's
/// hand-written `check_query_combinations` ladder.
///
/// Evaluation order: every `forbid` rule (table declaration order), THEN
/// every `atMostOne` rule (table declaration order) — see the module doc's
/// tie-break rationale. This never changes accept/reject, only which
/// `message` a multi-violation query gets.
pub fn check_query_combinations(present: &HashSet<&str>) -> Result<(), RuleViolation> {
    let forbid_violation = RULES
        .rules
        .iter()
        .filter_map(|rule| rule.forbid.as_ref().map(|clauses| (rule, clauses)))
        .find(|(_, clauses)| clauses.iter().all(|c| present.contains(c.as_str())));
    if let Some((rule, _)) = forbid_violation {
        return Err(RuleViolation {
            code: rule.code.clone(),
            message: rule.message.clone(),
        });
    }

    let at_most_one_violation = RULES
        .rules
        .iter()
        .filter_map(|rule| rule.at_most_one.as_ref().map(|clauses| (rule, clauses)))
        .find(|(_, clauses)| {
            clauses
                .iter()
                .filter(|c| present.contains(c.as_str()))
                .count()
                > 1
        });
    if let Some((rule, _)) = at_most_one_violation {
        return Err(RuleViolation {
            code: rule.code.clone(),
            message: rule.message.clone(),
        });
    }

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn table_parses_and_is_nonempty() {
        assert!(!RULES.rules.is_empty());
    }

    #[test]
    fn empty_query_has_no_violation() {
        let present: HashSet<&str> = HashSet::new();
        assert!(check_query_combinations(&present).is_ok());
    }

    #[test]
    fn get_and_index_together_is_rejected() {
        let present: HashSet<&str> = ["get", "index"].into_iter().collect();
        assert!(check_query_combinations(&present).is_err());
    }

    #[test]
    fn index_and_eq_together_is_allowed() {
        let present: HashSet<&str> = ["index", "eq"].into_iter().collect();
        assert!(check_query_combinations(&present).is_ok());
    }

    #[test]
    fn two_terminals_at_once_is_rejected() {
        let present: HashSet<&str> = ["count", "distinct"].into_iter().collect();
        assert!(check_query_combinations(&present).is_err());
    }
}
